File: developers/academy/py/standalone/which_search/_snippets/30_strategies.py
from uuid import UUID
import os
import weaviate
import weaviate.classes as wvc
import json


client = weaviate.connect_to_local(
    headers={
        "X-OpenAI-Api-Key": os.getenv("OPENAI_APIKEY")
    }
)


client.collections.delete("Product")

# START skipVectorizationExample
products = client.collections.create(
    name="Product",
    vectorizer_config=wvc.config.Configure.Vectorizer.text2vec_openai(
        # highlight-start
        vectorize_collection_name=True
        # highlight-end
        ),
    properties=[
        wvc.config.Property(
            name="name",
            data_type=wvc.config.DataType.TEXT,
            # highlight-start
            vectorize_property_name=True,
            # highlight-end
        ),
        wvc.config.Property(
            name="description",
            data_type=wvc.config.DataType.TEXT,
        ),
        wvc.config.Property(
            name="manufacturing_process",
            data_type=wvc.config.DataType.TEXT,
            # highlight-start
            skip_vectorization=True  # Skip unwanted property
            # highlight-end
        ),
    ]
)
# END skipVectorizationExample

client.collections.delete("Product")



client.collections.delete("SomeCollection")

# START tokenizationExample
things = client.collections.create(
    name="SomeCollection",
    properties=[
        wvc.config.Property(
            name="name",
            data_type=wvc.config.DataType.TEXT,
            # highlight-start
            tokenization=wvc.config.Tokenization.WORD  # Default
            # highlight-end
        ),
        wvc.config.Property(
            name="description",
            data_type=wvc.config.DataType.TEXT,
            tokenization=wvc.config.Tokenization.WHITESPACE  # Will keep case & special characters
        ),
        wvc.config.Property(
            name="email",
            data_type=wvc.config.DataType.TEXT,
            # highlight-start
            tokenization=wvc.config.Tokenization.FIELD  # Do not tokenize at all
            # highlight-end
        ),
    ]
)
# END tokenizationExample

client.collections.delete("SomeCollection")

from weaviate.classes.init import Auth

weaviate_url = os.getenv("WEAVIATE_URL")
weaviate_key = os.getenv("WEAVIATE_API_KEY")

client = weaviate.connect_to_weaviate_cloud(
    cluster_url=weaviate_url,
    auth_credentials=Auth.api_key(weaviate_key),
    headers={
        "X-OpenAI-Api-Key": os.getenv("OPENAI_APIKEY")
    }
)

# START selectAndBoostExample
questions = client.collections.get("JeopardyQuestion")

response = questions.query.bm25(
    "animal",
    limit=5,
    query_properties=["question^3", "answer"]  # Boost the impact of "question" property by 3
)

for o in response.objects:
    print(o.properties)
# END selectAndBoostExample


# START adjustAlpha
questions = client.collections.get("JeopardyQuestion")

response = questions.query.hybrid(
    "imaging",
    alpha=0.1,  # Mostly a vector search (Try it with alpha=0.9)
    limit=5
)

for o in response.objects:
    print(o.properties)
# END adjustAlpha


# START changeFusionType
questions = client.collections.get("JeopardyQuestion")

response = questions.query.hybrid(
    "imaging",
    fusion_type=wvc.query.HybridFusion.RELATIVE_SCORE,  # Use relative score fusion
    limit=5
)

for o in response.objects:
    print(o.properties)
# END changeFusionType

File: developers/academy/js/standalone/which-search/_snippets/30_strategies.ts
import weaviate, { CollectionConfig, CollectionConfigCreate, WeaviateClient, WeaviateReturn } from "weaviate-client"

let client: WeaviateClient

client = await weaviate.connectToWeaviateCloud(process.env.WEAVIATE_URL as string,{
    authCredentials: new weaviate.ApiKey(process.env.WEAVIATE_API_KEY as string),
    headers: {
      'X-OpenAI-Api-Key': process.env.OPENAI_APIKEY as string,  // Replace with your inference API key
    }
  }
)

client.collections.delete("Product")

// START skipVectorizationExample // START tokenizationExample
let products: CollectionConfigCreate
// END skipVectorizationExample // END tokenizationExample

// START skipVectorizationExample
products = await client.collections.create({
  name: "Product",
  vectorizers: weaviate.configure.vectorizer.text2VecOpenAI({
    // highlight-start
    vectorizeCollectionName: true
    // highlight-end
  }),
  properties: [{
    name: "name",
    dataType: weaviate.configure.dataType.TEXT,
    // highlight-start
    vectorizePropertyName: true
    // highlight-end
  },
  {
    name: "description",
    dataType: weaviate.configure.dataType.TEXT,
  },
  {
    name: "manufacturing_process",
    dataType: weaviate.configure.dataType.TEXT,
    // highlight-start
    skipVectorization: true,  // Skip unwanted property
    //highlight-end
  }]
})
// END skipVectorizationExample

client.collections.delete("Product")

client.collections.delete("SomeCollection")

// START tokenizationExample

products = await client.collections.create({
  name: "SomeCollection",
  properties: [{
    name: "name",
    dataType: weaviate.configure.dataType.TEXT,
    // highlight-start
    tokenization: weaviate.configure.tokenization.WORD // Default
    // highlight-end
  },
  {
    name: "description",
    dataType: weaviate.configure.dataType.TEXT,
    tokenization: weaviate.configure.tokenization.WHITESPACE // Will keep case & special characters
  },
  {
    name: "email",
    dataType: weaviate.configure.dataType.TEXT,
    // highlight-start
    tokenization: weaviate.configure.tokenization.FIELD // Do not tokenize at all
    // highlight-end
  }]
})
// END tokenizationExample


client.collections.delete("SomeCollection")

// START selectAndBoostExample // START adjustAlpha // START changeFusionType
type NonGenericReturn = WeaviateReturn<undefined>
let response: NonGenericReturn

const questions = client.collections.get("JeopardyQuestion")

// END selectAndBoostExample // END adjustAlpha // END changeFusionType


// START selectAndBoostExample
response = questions.query.bm25("animal",{ // Your query string
    limit: 5,
    queryProperties: ["question^3", "answer"] // Boost the impact of "question" property by 3
  }
)

for (const item of response.objects) {
  console.log(item.properties)
}
// END selectAndBoostExample

// START adjustAlpha
response = questions.query.hybrid("imaging",{ // Your query string
    limit: 5,
    alpha: 0.1, // Mostly a vector search (Try it with alpha=0.9)
  }
)

for (const item of response.objects) {
  console.log(item.properties)
}
// END adjustAlpha



// START changeFusionType
response = questions.query.hybrid("imaging",{ // Your query string
    limit: 5,
    fusionType: "RelativeScore",
    alpha: 0.1, // Mostly a vector search (Try it with alpha=0.9)
  }
)

for (const item of response.objects) {
  console.log(item.properties)
}
// END changeFusionType
